File: apps/staff/lib/customer-dossier.ts
/**
 * The Customer Dossier as the staff surface reads it (#635, #638; CONTEXT.md
 * "Customer Dossier", Ficha del cliente): everything one Event knows about one
 * Customer, addressed by the Customer's id and never by their email.
 *
 * SCOPED TO THE EVENT. The name and Tax ID arrive on each Sale because they are
 * what was given on THAT Sale — this Event's own snapshots — and never what the
 * platform-global Customer record says today. Nothing here joins or dedupes
 * them across Sales; two Sales with two Tax IDs show two Tax IDs.
 *
 * Later tickets widen `DossierSale` (#639) and add Tickets and Answers beside
 * `sales` (#640, #641). Pure apart from the fetch, so it runs under node --test.
 */

import { ApiError } from "./events-api.ts";
import {
  ASSIGNMENT_STATE_KEYS,
  assignmentStateBadgeVariant,
  type HolderStateKey,
  type TicketAssignmentState,
} from "./holder-list.ts";
import type { SaleTicketType } from "./sales-api.ts";

/*
  THE READ
*/

/**
 * A Sale's standing. `corrected` is a Sale reversed by a Sale Correction, told
 * apart by the API so this side need not infer it from the replacement ref.
 */
export type DossierSaleStatus = "active" | "reversed" | "corrected";

/** One of this Event's Ticket Sales to this Customer, as given on that Sale. */
export type DossierSale = {
  id: string;
  confirmation_ref: string;
  /** A string on the wire: narrow it with `saleStatusToken` before drawing. */
  status: string;
  /** When it was reversed; set for `reversed` and `corrected`. */
  reversed_at: string | null;
  /** The Sale Correction's replacement, on a `corrected` Sale. */
  replaced_by_confirmation_ref: string | null;
  sold_at: string;
  recorded_at: string;
  channel: string;
  source: string | null;
  origin: string;
  ticket_types: SaleTicketType[];
  amount_cents: number;
  currency: string;
  payment_method: string | null;
  customer_first_name: string;
  customer_last_name: string;
  tax_id_type: string | null;
  tax_id_number: string | null;
  /*
    What surrounds the Sale (#639); decided by `lib/dossier-sale-surroundings`.
  */
  /** The phone given on this Sale's checkout, off its Payment (ADR 0073); null when none was. */
  phone: string | null;
  /** The display name of the Affiliate Link that attributed the Sale, or null. */
  affiliate_link_name: string | null;
  /** The Sale's Tax Invoices in chain order; empty, never null. */
  tax_invoices: DossierTaxInvoice[];
  /** When the Sale was re-addressed (ADR 0058), or null. Never either address. */
  re_addressed_at: string | null;
  /*
    TICKETS AND HOLDERS (#640).
  */
  /** Every Ticket of the Sale; on a reversed or corrected Sale they are void. */
  tickets: DossierTicket[];
  /**
   * When the Assignment Reminder was sent for this Sale, oldest first. ABSENT
   * while `TICKET_ASSIGNMENT_ENABLED` is closed, so its absence is the flag.
   */
  assignment_reminder_sent_at?: string[];
};

/** One Tax Invoice about a Dossier Sale. Strings on the wire: narrow before drawing. */
export type DossierTaxInvoice = {
  kind: string;
  role: string;
  /** Null until signed. */
  number: string | null;
  status: string;
  recipient_legal_name: string;
  recipient_tax_id_type: string;
  recipient_tax_id: string;
};

export type CustomerDossier = {
  customer: { id: string; email: string };
  /** Newest `sold_at` first, as the API orders them. */
  sales: DossierSale[];
  /**
   * Tickets this Customer holds, accepted, on SOMEBODY ELSE's Sale (#640).
   * Absent while `TICKET_ASSIGNMENT_ENABLED` is closed.
   */
  held_tickets?: DossierHeldTicket[];
};

/**
 * One Ticket of one of the Customer's own Sales (#640).
 *
 * The assignment fields follow the Holder List's rules (ADR 0047): all absent
 * while the flag is closed and on a Sale that no longer stands; a Holder is
 * named only once they accepted; a purged assignment is `assigned` with
 * `never_accepted`. `self_held` marks the buyer's own Ticket (ADR 0048).
 */
export type DossierTicket = {
  ticket_id: string;
  ticket_type_id: string;
  ticket_type_name: string;
  ordinal: number;
  assignment_state?: TicketAssignmentState;
  never_accepted?: boolean;
  self_held?: boolean;
  holder_customer_id?: string;
  holder_first_name?: string;
  holder_last_name?: string;
};

/** A Ticket this Customer accepted on another buyer's Sale (#640). */
export type DossierHeldTicket = {
  ticket_id: string;
  ticket_type_id: string;
  ticket_type_name: string;
  ordinal: number;
  ticket_sale_id: string;
  confirmation_ref: string;
  /** `active` or `reversed` on the wire; read it with `heldTicketOnReversedSale`. */
  sale_status: string;
  buyer_first_name: string;
  buyer_last_name: string;
  accepted_at: string;
  /** The name this Customer gave as Holder. */
  holder_first_name?: string;
  holder_last_name?: string;
};

/**
 * What loading a Dossier can come to, short of a thrown failure.
 *
 * NOT FOUND IS AN OUTCOME AND NOT AN ERROR. The API answers 404 alike for an
 * Event outside the caller's Organization and for a Customer with no Sale on
 * this Event, and the page says one plain sentence for both. Read off the HTTP
 * status rather than an error code, so the sentence does not depend on which
 * `*_NOT_FOUND` code the API happens to choose.
 */
export type CustomerDossierResult =
  | { state: "found"; dossier: CustomerDossier }
  | { state: "not_found" };

/** Reads one Customer's Dossier on one Event through the BFF route. */
export async function fetchCustomerDossier(
  eventId: string,
  customerId: string,
): Promise<CustomerDossierResult> {
  const response = await fetch(
    `/api/events/${encodeURIComponent(eventId)}/customers/${encodeURIComponent(customerId)}/dossier`,
    { headers: { "Content-Type": "application/json" } },
  );
  if (response.status === 404) {
    return { state: "not_found" };
  }
  const envelope = (await response.json().catch(() => null)) as {
    data?: CustomerDossier | null;
    error?: { message?: string; code?: string; details?: Record<string, unknown> } | null;
  } | null;
  if (!response.ok || envelope?.error || !envelope?.data) {
    throw new ApiError(
      envelope?.error?.message ?? "",
      envelope?.error?.code,
      envelope?.error?.details,
    );
  }
  return { state: "found", dossier: envelope.data };
}

/*
  THE ADDRESS, AND THE WAY BACK
*/

/** The Sales list: where Back goes when `from` cannot be trusted. */
function salesListPath(eventId: string): string {
  return `/events/${encodeURIComponent(eventId)}/sales`;
}

/**
 * The Dossier's own address. The Customer id is the only thing naming the
 * person, so no email reaches history, a pasted link or a log (#635 story 8).
 * `from` is the list the reader came from, filters and all, for Back.
 */
export function dossierHref(eventId: string, customerId: string, from: string): string {
  return `${salesListPath(eventId)}/customers/${encodeURIComponent(customerId)}?from=${encodeURIComponent(from)}`;
}

// A dot segment, literal or percent-encoded: a URL parser resolves `%2e%2e`
// exactly as it resolves `..`, so both would walk out of the Event.
const DOT_SEGMENT = /^(?:\.|%2e){1,2}$/i;
// C0 controls and DEL. A browser strips tabs and newlines from a URL, which is
// how `/\t/evil.example` becomes `//evil.example`.
const CONTROL_CHARACTER = /[\u0000-\u001f\u007f]/;

/**
 * Where Back goes: `from`, when it is a relative path within this Event's own
 * staff pages, and the Sales list otherwise.
 *
 * `from` is whatever a URL said, so it is an OPEN REDIRECT unless held to this:
 * it must begin `/events/<eventId>` and continue with `/`, `?`, `#` or nothing,
 * which rules out absolute URLs, schemes, protocol-relative `//host`, another
 * Event and an Event whose id merely starts with this one's. Backslashes and
 * control characters are refused because browsers rewrite them into slashes or
 * drop them, and dot segments because they climb out of the prefix after it has
 * been checked. `from` arrives already decoded from the search params.
 */
export function dossierBackHref(from: string | null | undefined, eventId: string): string {
  const fallback = salesListPath(eventId);
  if (typeof from !== "string" || from === "" || eventId === "") {
    return fallback;
  }
  if (from.includes("\\") || CONTROL_CHARACTER.test(from)) {
    return fallback;
  }
  const prefix = `/events/${encodeURIComponent(eventId)}`;
  if (!from.startsWith(prefix)) {
    return fallback;
  }
  const next = from.charAt(prefix.length);
  if (next !== "" && next !== "/" && next !== "?" && next !== "#") {
    return fallback;
  }
  const path = from.split(/[?#]/, 1)[0];
  if (path.split("/").some((segment) => DOT_SEGMENT.test(segment))) {
    return fallback;
  }
  return from;
}

/*
  LABELLING
*/

export const DOSSIER_SALE_STATUSES = ["active", "reversed", "corrected"] as const satisfies readonly DossierSaleStatus[];

/**
 * A Sale's status as a token this app has words for, or null for one it has
 * never heard of — the caller then shows the API's own word rather than
 * guessing, the same floor `saleChannelToken` keeps (ADR 0041).
 */
export function saleStatusToken(status: string | null | undefined): DossierSaleStatus | null {
  return status != null && (DOSSIER_SALE_STATUSES as readonly string[]).includes(status)
    ? (status as DossierSaleStatus)
    : null;
}

/**
 * The Badge variant a status is drawn in. A Sale that no longer stands is
 * destructive, as on the sales list; an unrecognised status is drawn quietly
 * rather than as a live Sale, because nothing says it is one.
 */
export function saleStatusBadgeVariant(
  token: DossierSaleStatus | null,
): "secondary" | "destructive" | "outline" {
  switch (token) {
    case "active":
      return "secondary";
    case "reversed":
    case "corrected":
      return "destructive";
    default:
      return "outline";
  }
}

/**
 * The name given on one Sale, the two halves joined, or "" when neither was
 * given — so the caller draws its "nothing to show" rather than a stray space.
 */
export function nameGivenOnSale(sale: Pick<DossierSale, "customer_first_name" | "customer_last_name">): string {
  return [sale.customer_first_name, sale.customer_last_name]
    .map((part) => (part ?? "").trim())
    .filter(Boolean)
    .join(" ");
}

/*
  TICKETS AND HOLDERS (#640)
*/

function joinName(first: string | undefined, last: string | undefined): string {
  return [first ?? "", last ?? ""]
    .map((part) => part.trim())
    .filter(Boolean)
    .join(" ");
}

/**
 * The words one Ticket's assignment is named with: the Holder List's own
 * `outstandingAnswers` keys, so the two screens cannot describe one Ticket
 * differently — plus `selfHeld`, a `customerDossier` key, for the buyer's own
 * Ticket, which is theirs and not given away. Null when the payload carries no
 * state (the flag closed, or a void Sale): there is nothing to draw.
 */
export type DossierTicketStateKey = HolderStateKey | "selfHeld";

export function dossierTicketStateKey(ticket: DossierTicket): DossierTicketStateKey | null {
  if (!ticket.assignment_state) {
    return null;
  }
  if (ticket.self_held) {
    return "selfHeld";
  }
  if (ticket.never_accepted) {
    return "holderNeverAccepted";
  }
  return ASSIGNMENT_STATE_KEYS[ticket.assignment_state];
}

/** The Badge variant for `dossierTicketStateKey`, in the Holder List's colours. */
export function dossierTicketBadgeVariant(
  ticket: DossierTicket,
): "success" | "warning" | "outline" | "secondary" | null {
  if (!ticket.assignment_state) {
    return null;
  }
  if (ticket.self_held) {
    return "secondary";
  }
  if (ticket.never_accepted) {
    return "outline";
  }
  return assignmentStateBadgeVariant(ticket.assignment_state);
}

/** The accepted Holder's name on a Ticket, or "" when nobody is named. */
export function holderNameOnTicket(ticket: DossierTicket): string {
  return joinName(ticket.holder_first_name, ticket.holder_last_name);
}

/**
 * The Customer whose Dossier a Ticket's Holder name opens, or null for no link:
 * only an accepted Holder, and never the buyer's own Self-held Ticket, which
 * would only link back to the page it is on.
 */
export function ticketHolderDossierId(ticket: DossierTicket): string | null {
  if (ticket.assignment_state !== "accepted" || ticket.self_held || !ticket.holder_customer_id) {
    return null;
  }
  return ticket.holder_customer_id;
}

/** Whether a Sale's Assignment Reminder times exist to draw; off the payload's absence. */
export function assignmentRemindersVisible(
  sale: Pick<DossierSale, "assignment_reminder_sent_at">,
): boolean {
  return sale.assignment_reminder_sent_at !== undefined;
}

/** Whether the held Tickets section exists; off the payload's absence (ADR 0045). */
export function heldTicketsVisible(dossier: Pick<CustomerDossier, "held_tickets">): boolean {
  return dossier.held_tickets !== undefined;
}

/** The buyer of the Sale a held Ticket belongs to, or "" when no name was given. */
export function heldTicketBuyerName(ticket: DossierHeldTicket): string {
  return joinName(ticket.buyer_first_name, ticket.buyer_last_name);
}

/** The name this Customer gave as the Holder, or "" when none was given. */
export function nameGivenAsHolder(ticket: DossierHeldTicket): string {
  return joinName(ticket.holder_first_name, ticket.holder_last_name);
}

/**
 * Whether the held Ticket's Sale no longer stands, so there is no live Ticket.
 * Anything but `active` counts, so an unfamiliar status is never drawn as live.
 */
export function heldTicketOnReversedSale(ticket: DossierHeldTicket): boolean {
  return ticket.sale_status !== "active";
}

File: backend/internal/catalog/service/customer_dossier.go
package service

import (
	"context"
	"time"

	"github.com/peter/ticket_pos/backend/internal/catalog"
	"github.com/peter/ticket_pos/backend/internal/catalog/repository"
	"github.com/peter/ticket_pos/backend/internal/sales"
)

// The Customer Dossier (#638, spec #635; CONTEXT.md "Customer Dossier"): the
// one read of everything an Event knows about one Customer.
//
// ONE CALL, ONE ASSEMBLY. GetCustomerDossier is the whole interface: the
// handler asks for a Dossier and gets a finished one. Each later part of the
// Dossier — a Sale's phone, Affiliate Link, invoices and re-addressing (#639),
// its Tickets and the Tickets held on other Sales (#640), their Answers (#641)
// — is a field on CustomerDossier or DossierSale, filled here, so no caller
// ever composes a Dossier out of pieces.
//
// SCOPED TO THE EVENT, NEVER TO THE PLATFORM-GLOBAL CUSTOMER. From the Customer
// record it takes the id and the email identity; every name and Tax ID comes
// off this Event's own Sales, beside the Sale it was given on. A later purchase
// at another Organization rewrites nothing shown here.
//
// NOTHING AT THIS EVENT IS NOT FOUND. A Customer with no Ticket Sale on this
// Event is answered exactly as an id that names nobody, so the Dossier cannot
// be used to learn whether somebody bought elsewhere. #640 widens "something at
// this Event" to a Ticket held on another buyer's Sale.

// Dossier Sale statuses. `corrected` is a reversed Sale that a Sale Correction
// replaced — the same reading the Sales list gives a reversed row naming its
// replacement — and `upgraded` is a reversed Sale its own buyer surrendered for
// a paid one.
//
// THE LAST TWO ARE BOTH "REVERSED AND REPLACED" AND THEY ARE STILL TWO WORDS
// (#651, ADR 0074). `corrected` is ADR 0050's, and it says somebody recorded a
// sale wrongly and staff put it right; an Upgrade is a Sale recorded perfectly
// by a buyer who changed their mind, and no member of staff ever touched it.
// One word for both would tell an Organization its people erred on a Sale nobody
// there has ever seen.
const (
	DossierSaleActive    = "active"
	DossierSaleReversed  = "reversed"
	DossierSaleCorrected = "corrected"
	DossierSaleUpgraded  = "upgraded"
)

// CustomerDossier is the Customer Dossier response.
type CustomerDossier struct {
	Customer DossierCustomerView `json:"customer"`
	// Sales are this Event's Ticket Sales to the Customer, reversed and
	// corrected ones included, newest first.
	Sales []DossierSaleView `json:"sales"`
	// HeldTickets are the Tickets of this Event the Customer ACCEPTED on
	// somebody else's Sale, reversed Sales included, newest Sale first (#640).
	// Absent while TICKET_ASSIGNMENT_ENABLED is closed (ADR 0045); `[]` when
	// they hold none.
	HeldTickets *[]DossierHeldTicketView `json:"held_tickets,omitempty"`
}

// DossierCustomerView is who the Customer is, platform-wide: an id and an
// email, and deliberately nothing more.
type DossierCustomerView struct {
	ID    string `json:"id"`
	Email string `json:"email"`
}

// DossierSaleView is one of this Event's Ticket Sales on the Dossier, with the
// name and Tax ID given on it.
type DossierSaleView struct {
	ID              string `json:"id"`
	ConfirmationRef string `json:"confirmation_ref"`
	// Status is `active`, `reversed`, `corrected` or `upgraded`.
	Status string `json:"status" enums:"active,reversed,corrected,upgraded"`
	// ReversedAt is when a reversed, corrected or upgraded Sale was reversed;
	// null on an active one.
	ReversedAt *time.Time `json:"reversed_at"`
	// ReplacedByConfirmationRef names the replacement of a corrected Sale, or
	// the paid Sale an upgraded one was surrendered for.
	ReplacedByConfirmationRef *string `json:"replaced_by_confirmation_ref"`
	// ReplacesConfirmationRef is the other half of that pair, stated on the
	// REPLACEMENT: the mistaken Sale it corrects, or the free Sale its buyer
	// gave up for it. Both halves are published so an Organization reading
	// either row can explain to a buyer where a Ticket went (#651).
	ReplacesConfirmationRef *string `json:"replaces_confirmation_ref"`
	// ReplacementReason is WHY the two Sales are linked: `correction` or
	// `upgrade`, null when there is no link.
	//
	// IT IS FOR THE REPLACEMENT'S HALF, which `status` cannot speak for. The
	// replaced Sale's word arrives in `status` — `corrected` or `upgraded`,
	// decided here so the page never infers it — but the Sale standing in its
	// place is simply `active`, and it still has to say whether it corrects a
	// mistake or stands in for a Ticket its buyer traded up from.
	ReplacementReason *string   `json:"replacement_reason" enums:"correction,upgrade"`
	SoldAt            time.Time `json:"sold_at"`
	RecordedAt        time.Time `json:"recorded_at"`
	// Channel is the Sales Channel: `online`, `in_person` or `import`.
	Channel string `json:"channel"`
	// Source is the Sales list's `source` for the Sale.
	Source *string `json:"source"`
	// Origin is how the Sale reached the platform, derived exactly as the Sales
	// list derives it (ADR 0052).
	Origin            string                `json:"origin"`
	TicketTypes       []DossierSaleLineView `json:"ticket_types"`
	AmountCents       int                   `json:"amount_cents"`
	Currency          string                `json:"currency"`
	PaymentMethod     *string               `json:"payment_method"`
	CustomerFirstName string                `json:"customer_first_name"`
	CustomerLastName  string                `json:"customer_last_name"`
	TaxIDType         *string               `json:"tax_id_type"`
	TaxIDNumber       *string               `json:"tax_id_number"`

	// What surrounds the Sale (#639) — see customer_dossier_surroundings.go.

	// Phone is the number given on the checkout this Sale came from, read off
	// its Payment (ADR 0073) and never the Customer record's current phone.
	// Null for a Sale with no checkout behind it — In-Person, imported or
	// manually recorded — or a checkout that gave none.
	Phone *string `json:"phone"`
	// AffiliateLinkName is the display name of the Affiliate Link that
	// attributed the Sale, or null.
	AffiliateLinkName *string `json:"affiliate_link_name"`
	// TaxInvoices are the Tax Invoices about the Sale — its Sale Invoices and
	// any Credit Notes — as the operator Sale lookup
	// reads them, in the Sale's chain order; empty, never null.
	TaxInvoices []DossierTaxInvoiceView `json:"tax_invoices"`
	// ReAddressedAt is when the Sale was re-addressed (ADR 0058) — the
	// acceptance of its latest accepted Sale Re-addressing — or null. Neither
	// address and no token is ever carried.
	ReAddressedAt *time.Time `json:"re_addressed_at"`
	// Tickets and assignment reminders (#640).

	// Tickets are the Sale's Tickets in the catalog's order. A reversed or
	// corrected Sale's Tickets are void and carry their Ticket Type, ordinal
	// and Answers only; `status` is what says so.
	Tickets []DossierTicketView `json:"tickets"`
	// AssignmentReminderSentAt is when an Assignment Reminder was sent about
	// this Sale, oldest first. Absent while TICKET_ASSIGNMENT_ENABLED is closed.
	AssignmentReminderSentAt *[]time.Time `json:"assignment_reminder_sent_at,omitempty"`
}

// DossierTicketView is one Ticket of a Dossier Sale.
//
// EVERY ASSIGNMENT FIELD COMES FROM catalog.DiscloseHolder (ADR 0047) and is
// `omitempty`: absent while TICKET_ASSIGNMENT_ENABLED is closed and on a
// reversed Sale. An unaccepted assignment is a state and never a person.
type DossierTicketView struct {
	// TicketID is what the Ticket's Answers are read and given by.
	TicketID       string `json:"ticket_id"`
	TicketTypeID   string `json:"ticket_type_id"`
	TicketTypeName string `json:"ticket_type_name"`
	// Ordinal is which of its Ticket Sale Line's units this Ticket is.
	Ordinal int `json:"ordinal"`
	// AssignmentState is `unassigned`, `assigned` or `accepted`.
	AssignmentState string `json:"assignment_state,omitempty" enums:"unassigned,assigned,accepted"`
	// NeverAccepted marks an assignment whose unaccepted address the retention
	// purge took; it reads `assigned` beside it (#334).
	NeverAccepted bool `json:"never_accepted,omitempty"`
	// SelfHeld marks a Ticket this Customer accepted on their own Sale — their
	// Self-held Ticket (ADR 0048) — which names nobody else.
	SelfHeld bool `json:"self_held,omitempty"`
	// The accepted Holder when that is somebody else: their Customer id, which
	// links their Dossier, and the name they gave as Holder.
	HolderCustomerID string `json:"holder_customer_id,omitempty"`
	HolderFirstName  string `json:"holder_first_name,omitempty"`
	HolderLastName   string `json:"holder_last_name,omitempty"`
	// Answers, Outstanding Answers and the last Answer Reminder (#641); a
	// Ticket on a reversed Sale carries its Answers only.
	DossierTicketAnswers
}

// DossierHeldTicketView is a Ticket the Customer holds on somebody else's Sale.
type DossierHeldTicketView struct {
	TicketID       string `json:"ticket_id"`
	TicketTypeID   string `json:"ticket_type_id"`
	TicketTypeName string `json:"ticket_type_name"`
	Ordinal        int    `json:"ordinal"`
	// TicketSaleID and ConfirmationRef name the Sale it is on.
	TicketSaleID    string `json:"ticket_sale_id"`
	ConfirmationRef string `json:"confirmation_ref"`
	// SaleStatus is `reversed` when that Sale was reversed (a Sale Correction
	// included): the Ticket is void, and this Customer held it.
	SaleStatus string `json:"sale_status" enums:"active,reversed"`
	// The buyer's name as given on that Sale.
	BuyerFirstName string `json:"buyer_first_name"`
	BuyerLastName  string `json:"buyer_last_name"`
	// AcceptedAt is when this Customer accepted the Ticket.
	AcceptedAt *time.Time `json:"accepted_at"`
	// The name this Customer gave as its Holder.
	HolderFirstName string `json:"holder_first_name,omitempty"`
	HolderLastName  string `json:"holder_last_name,omitempty"`
	// Answers, Outstanding Answers and the last Answer Reminder (#641); a
	// Ticket on a reversed Sale carries its Answers only.
	DossierTicketAnswers
}

// DossierSaleLineView is one Ticket Type bought on a Dossier Sale.
type DossierSaleLineView struct {
	TicketTypeID   string `json:"ticket_type_id"`
	TicketTypeName string `json:"ticket_type_name"`
	Quantity       int    `json:"quantity"`
}

// GetCustomerDossier assembles the Customer Dossier of one Customer at one of
// the caller's Organization's Events.
func (s *Service) GetCustomerDossier(
	ctx context.Context, actor ActorContext, eventID, customerID string,
) (*CustomerDossier, error) {
	event, err := s.repo.GetEventByID(ctx, actor.OrganizationID, eventID)
	if err != nil {
		return nil, err
	}
	if event == nil {
		return nil, catalog.ErrEventNotFound()
	}

	saleRows, err := s.repo.ListDossierSales(ctx, actor.OrganizationID, eventID, customerID)
	if err != nil {
		return nil, err
	}
	// #640: a Customer who holds a Ticket of this Event — accepted, on somebody
	// else's Sale, reversed or not — is found too; never while assignment is dark.
	var heldTickets []DossierHeldTicketView
	if s.ticketAssignmentEnabled {
		if heldTickets, err = s.dossierHeldTickets(ctx, actor, eventID, customerID); err != nil {
			return nil, err
		}
	}
	if len(saleRows) == 0 && len(heldTickets) == 0 {
		return nil, catalog.ErrCustomerNotFoundAtEvent()
	}
	customer, err := s.repo.GetDossierCustomer(ctx, customerID)
	if err != nil {
		return nil, err
	}
	if customer == nil {
		return nil, catalog.ErrCustomerNotFoundAtEvent()
	}

	dossier := &CustomerDossier{
		Customer: DossierCustomerView{ID: customer.ID, Email: customer.Email},
		Sales:    make([]DossierSaleView, 0, len(saleRows)),
	}
	for _, row := range saleRows {
		dossier.Sales = append(dossier.Sales, dossierSaleView(row))
	}
	if err := s.fillDossierSaleSurroundings(ctx, actor, eventID, dossier.Sales); err != nil {
		return nil, err
	}
	if err := s.fillDossierTickets(ctx, actor, eventID, customerID, dossier.Sales); err != nil {
		return nil, err
	}
	if err := s.fillDossierAnswers(ctx, actor, eventID, dossier.Sales, heldTickets); err != nil {
		return nil, err
	}
	if s.ticketAssignmentEnabled {
		if heldTickets == nil {
			heldTickets = []DossierHeldTicketView{}
		}
		dossier.HeldTickets = &heldTickets
	}
	return dossier, nil
}

// fillDossierTickets puts each Sale's Tickets and, while assignment is open, its
// assignment reminder send times on the Sales (#640).
func (s *Service) fillDossierTickets(
	ctx context.Context, actor ActorContext, eventID, customerID string, sales []DossierSaleView,
) error {
	saleIDs := make([]string, 0, len(sales))
	for _, sale := range sales {
		saleIDs = append(saleIDs, sale.ID)
	}
	tickets, err := s.repo.ListDossierSaleTickets(ctx, actor.OrganizationID, eventID, saleIDs)
	if err != nil {
		return err
	}
	ticketsBySale := make(map[string][]repository.DossierTicket, len(sales))
	for _, ticket := range tickets {
		ticketsBySale[ticket.TicketSaleID] = append(ticketsBySale[ticket.TicketSaleID], ticket)
	}

	remindersBySale := map[string][]time.Time{}
	if s.ticketAssignmentEnabled {
		reminders, err := s.repo.ListDossierAssignmentReminders(ctx, actor.OrganizationID, eventID, saleIDs)
		if err != nil {
			return err
		}
		for _, reminder := range reminders {
			remindersBySale[reminder.TicketSaleID] = append(remindersBySale[reminder.TicketSaleID], reminder.SentAt)
		}
	}

	for i := range sales {
		sale := &sales[i]
		live := sale.Status == DossierSaleActive
		sale.Tickets = make([]DossierTicketView, 0, len(ticketsBySale[sale.ID]))
		for _, ticket := range ticketsBySale[sale.ID] {
			view := DossierTicketView{
				TicketID:       ticket.ID,
				TicketTypeID:   ticket.TicketTypeID,
				TicketTypeName: ticket.TicketTypeName,
				Ordinal:        ticket.Ordinal,
			}
			// A reversed Sale's Tickets are void: no live assignment state.
			if live {
				fillDossierTicketHolder(&view, s.discloseDossierHolder(ticket), customerID)
			}
			sale.Tickets = append(sale.Tickets, view)
		}
		if s.ticketAssignmentEnabled {
			sent := remindersBySale[sale.ID]
			if sent == nil {
				sent = []time.Time{}
			}
			sale.AssignmentReminderSentAt = &sent
		}
	}
	return nil
}

// fillDossierTicketHolder shapes one live Ticket's disclosure. The buyer's own
// acceptance is their Self-held Ticket and names nobody; anybody else's is
// named and linked. A closed flag discloses nothing, so nothing is filled.
func fillDossierTicketHolder(view *DossierTicketView, disclosure catalog.HolderDisclosure, customerID string) {
	view.AssignmentState = string(disclosure.State)
	view.NeverAccepted = disclosure.NeverAccepted
	if disclosure.HolderCustomerID == "" {
		return
	}
	if disclosure.HolderCustomerID == customerID {
		view.SelfHeld = true
		return
	}
	view.HolderCustomerID = disclosure.HolderCustomerID
	view.HolderFirstName = disclosure.HolderFirstName
	view.HolderLastName = disclosure.HolderLastName
}

// dossierHeldTickets returns the Tickets the Customer holds on other buyers'
// Sales of this Event. The query selects accepted Tickets naming the Customer,
// and each is kept only if DiscloseHolder discloses that very Customer, so the
// filter can never find somebody the rule would not name.
func (s *Service) dossierHeldTickets(
	ctx context.Context, actor ActorContext, eventID, customerID string,
) ([]DossierHeldTicketView, error) {
	rows, err := s.repo.ListDossierHeldTickets(ctx, actor.OrganizationID, eventID, customerID)
	if err != nil {
		return nil, err
	}
	views := make([]DossierHeldTicketView, 0, len(rows))
	for _, row := range rows {
		disclosure := s.discloseDossierHolder(row.DossierTicket)
		if disclosure.HolderCustomerID != customerID {
			continue
		}
		saleStatus := DossierSaleActive
		if row.SaleStatus == DossierSaleReversed {
			saleStatus = DossierSaleReversed
		}
		views = append(views, DossierHeldTicketView{
			TicketID:        row.ID,
			TicketTypeID:    row.TicketTypeID,
			TicketTypeName:  row.TicketTypeName,
			Ordinal:         row.Ordinal,
			TicketSaleID:    row.TicketSaleID,
			ConfirmationRef: row.ConfirmationRef,
			SaleStatus:      saleStatus,
			BuyerFirstName:  row.BuyerFirstName,
			BuyerLastName:   row.BuyerLastName,
			AcceptedAt:      disclosure.AcceptedAt,
			HolderFirstName: disclosure.HolderFirstName,
			HolderLastName:  disclosure.HolderLastName,
		})
	}
	return views, nil
}

// discloseDossierHolder is the Dossier's adapter onto catalog.DiscloseHolder,
// as fillHolderListEntry is the Holder List's.
func (s *Service) discloseDossierHolder(ticket repository.DossierTicket) catalog.HolderDisclosure {
	return catalog.DiscloseHolder(s.ticketAssignmentEnabled, catalog.HolderAssignment{
		HolderEmail:           ticket.HolderEmail.String,
		HolderFirstName:       ticket.HolderFirstName.String,
		HolderLastName:        ticket.HolderLastName.String,
		HolderCustomerID:      ticket.HolderCustomerID.String,
		AssignedAt:            nullTimeOrNil(ticket.AssignedAt),
		AcceptedAt:            nullTimeOrNil(ticket.AcceptedAt),
		HolderAddressPurgedAt: nullTimeOrNil(ticket.HolderAddressPurgedAt),
	})
}

// dossierSaleView shapes one Sale row for the Dossier.
func dossierSaleView(row repository.DossierSale) DossierSaleView {
	lines := make([]DossierSaleLineView, 0, len(row.TicketTypes))
	for _, line := range row.TicketTypes {
		lines = append(lines, DossierSaleLineView(line))
	}
	return DossierSaleView{
		ID:                        row.ID,
		ConfirmationRef:           row.ConfirmationRef,
		Status:                    dossierSaleStatus(row),
		ReversedAt:                row.ReversedAt,
		ReplacedByConfirmationRef: row.ReplacedByConfirmationRef,
		ReplacesConfirmationRef:   row.ReplacesConfirmationRef,
		ReplacementReason:         row.ReplacementReason,
		SoldAt:                    row.SoldAt,
		RecordedAt:                row.RecordedAt,
		Channel:                   row.Channel,
		Source:                    row.Source,
		Origin:                    sales.DeriveSaleOrigin(row.Channel, row.ImportBatchID, row.ReplacesSaleID),
		TicketTypes:               lines,
		AmountCents:               row.AmountCents,
		Currency:                  row.Currency,
		PaymentMethod:             row.PaymentMethod,
		CustomerFirstName:         row.CustomerFirstName,
		CustomerLastName:          row.CustomerLastName,
		TaxIDType:                 row.TaxIDType,
		TaxIDNumber:               row.TaxIDNumber,
	}
}

// dossierSaleStatus reads a Sale's status for the Dossier: a reversed Sale a
// correction replaced is `corrected`, one its buyer upgraded out of is
// `upgraded`, and a reversal with no replacement behind it is just `reversed`.
//
// THE REASON IS READ AND THE LINK IS NOT ENOUGH (#651, ADR 0074). Until the
// Upgrade there was one way to replace a Ticket Sale, so the link alone could
// stand for the reason; there are two now, and `replacement_reason` is the
// column that tells them apart (migration 123, written on both halves so no join
// is needed here).
//
// ONLY THE UPGRADE'S OWN WORD MOVES THIS OFF `corrected`. The fallthrough is
// deliberate rather than defensive: a link whose reason this build cannot read —
// a third reason added later, or a row written by an older binary — keeps the
// older, narrower sentence instead of being promoted to a claim about a buyer's
// election that nobody made.
func dossierSaleStatus(row repository.DossierSale) string {
	if row.Status != DossierSaleReversed {
		return DossierSaleActive
	}
	if row.ReplacedBySaleID == nil {
		return DossierSaleReversed
	}
	if sales.IsUpgradeReplacement(row.ReplacementReason) {
		return DossierSaleUpgraded
	}
	return DossierSaleCorrected
}
